test/xbgpu/test_engine.py

```diff
@@ -1201,7 +1212,7 @@ def heap_factory(batch_index: int, present: np.ndarray) -> list[spead2.send.Heap
 
         expected_beam_saturated_low, expected_beam_saturated_high = verify_beam_data(
             beam_outputs=beam_outputs,
-            beam_results=beam_results,
+            beam_results=list(beam_results.values()),
```

Comment:
This makes me slightly nervous: it depends on the iteration order of the beam_results dictionary matching the order of `beam_outputs`. That is guaranteed with the way you've written the code, but it's an implicit assumption that someone might break then spend time debugging. Some alternative ideas:

1. 
```suggestion
            beam_results=[beam_results[output] for output in beam_outputs],
```

2. Change the parameter in verify_beam_data to a dictionary.
3. Add a comment where `beam_results` is defined to indicate that the iteration order is significant.

Some comment applies to the verify_corrprod_data call.

Reply:
Noted all around and appreciate you flagging this. Must admit, I had similar concerns, but hadn't considered this particular case.

Regarding
> 2. Change the parameter in verify_beam_data to a dictionary

I'm not quite seeing why that needs to be done if you've suggested list comprehension for the `beam_results` values?
* That is, `[my_dict[this_key] for this_key in my_keys]` just results in a list of values (at least in how I've tested the logic in an ipython session)

Reply:
I meant do one of those things, not all of them.